tests/test_grammar.py

```diff
@@ -309,6 +308,7 @@
     ("str_top_middle_escapes", r"abc\\\\\${str}", r"abc\\${str}"),
     ("str_top_trailing_escapes", "${str}" + "\\" * 5, "hi" + "\\" * 3),
     ("str_top_concat_interpolations", "${null}${float}", "None1.2"),
+    ("str_top_issue_617", r""" ${test: "hi\\" }"} """, ' hi\\"} '),
```

Comment:
I would like to see many more less confusing tests around this issue to convince me this is working in an intuitive way.

Ideally, those tests would be in a standalone test function and not overload the already big config object.

Reply:
Agreed on more tests, I realized it afterwards and thought I'd bring in some tests from #621 before merging. I thought I'd wait until I'd get some feedback on the issue + fix first. I guess now is the time :)

There's no need to change the config itself since it's just about testing quoted strings => done in cd17736
If you're curious, here are the tests that would fail before this PR:
```
=========================================================================== short test summary info ===========================================================================
FAILED tests/test_grammar.py::TestOmegaConfGrammar::test_single_element_no_interpolation[str_quoted_bad_1] - Failed: DID NOT RAISE <class 'omegaconf.errors.GrammarParseError'>
FAILED tests/test_grammar.py::TestOmegaConfGrammar::test_single_element_no_interpolation[str_quoted_bad_2] - Failed: DID NOT RAISE <class 'omegaconf.errors.GrammarParseError'>
FAILED tests/test_grammar.py::TestOmegaConfGrammar::test_single_element_no_interpolation[str_quoted_esc_quote_single_bad] - Failed: DID NOT RAISE <class 'omegaconf.errors.G...
FAILED tests/test_grammar.py::TestOmegaConfGrammar::test_single_element_no_interpolation[str_quoted_esc_quote_double_bad] - Failed: DID NOT RAISE <class 'omegaconf.errors.G...
FAILED tests/test_grammar.py::TestOmegaConfGrammar::test_config_value[str_top_issue_617] - assert ' hi\\" } ' == ' hi\\"} '
======================================================================== 5 failed, 248 passed in 0.78s ========================================================================
```